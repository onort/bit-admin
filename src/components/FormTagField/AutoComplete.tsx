import React, { PureComponent } from "react"
import { ApolloConsumer } from "react-apollo"
import { ApolloClient, ApolloQueryResult } from "apollo-boost"
import gql from "graphql-tag"
import AutoSuggest, {
  ChangeEvent,
  SuggestionSelectedEventData
} from "react-autosuggest"

import styles from "./FormTagField.scss"
import { TagData } from "../../types"

const tagQuery = gql`
  query Tag($query: String!) {
    getTags(query: $query) {
      name
      id
    }
  }
`

interface Props {
  fieldName: string
  onSelect: (suggestion: TagData) => void
  setTouched: (field: string) => void
}

interface State {
  value: string
  suggestions: TagData[]
}

class AutoComplete extends PureComponent<Props, State> {
  public state = { value: "", suggestions: [] }

  public onChange = (e: React.FormEvent<any>, { newValue }: ChangeEvent) => {
    this.setState({
      value: newValue
    })
  }

  public onBlur = (e: React.FormEvent<any>) => {
    this.props.setTouched(this.props.fieldName)
  }

  public onSuggestionsFetchRequested = (client: ApolloClient<any>) => ({
    value
  }: {
    value: string
  }) => {
    client
      .query({
        query: tagQuery,
        variables: { query: value }
      })
      .then(({ data }: ApolloQueryResult<any>) => {
        this.setState({ suggestions: data.getTags })
      })
  }

  public onSuggestionsClearRequested = () => {
    this.setState({ value: "", suggestions: [] })
  }

  public onSuggestionSelected = (
    e: React.FormEvent<any>,
    { suggestion }: SuggestionSelectedEventData<any>
  ) => {
    e.preventDefault()
    this.props.onSelect(suggestion)
  }

  public getSuggestionValue = (suggestion: TagData) => suggestion.name

  public renderSuggestion = (suggestion: TagData) => (
    <div className={styles.suggestionName}>
      &#35;
      {suggestion.name}
    </div>
  )

  public render() {
    const { value, suggestions } = this.state

    const inputProps = {
      placeholder: "Select at least a tag",
      value,
      onBlur: this.onBlur,
      onChange: this.onChange,
      autoCorrect: "off",
      autoCapitalize: "off",
      autoComplete: "off"
    }

    return (
      <ApolloConsumer>
        {(client: ApolloClient<any>) => (
          <AutoSuggest
            suggestions={suggestions}
            onSuggestionsFetchRequested={this.onSuggestionsFetchRequested(
              client
            )}
            onSuggestionsClearRequested={this.onSuggestionsClearRequested}
            onSuggestionSelected={this.onSuggestionSelected}
            getSuggestionValue={this.getSuggestionValue}
            renderSuggestion={this.renderSuggestion}
            inputProps={inputProps}
            theme={{
              input: styles.input,
              container: styles.autoCompleteContainer,
              suggestionsContainerOpen: styles.suggestionsContainerOpen,
              suggestionsList: styles.suggestionsList,
              suggestion: styles.suggestion,
              suggestionHighlighted: styles.suggestionHighlighted
            }}
          />
        )}
      </ApolloConsumer>
    )
  }
}

export default AutoComplete
