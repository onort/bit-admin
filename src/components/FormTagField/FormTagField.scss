@import "../../styles/main.scss";

.container {
  width: 100%;
  margin: $spacer-sm 0;
}

.label {
  display: block;
  font-size: $font-size-md;
  font-weight: 600;
  letter-spacing: 0.66px;
  margin-bottom: $spacer-xxsm;
}

.input {
  display: block;
  width: 100%;
  border: 1px solid rgba(34, 36, 38, 0.15);
  background-color: $white;
  border-radius: px-to-rem(4px);
  padding: $spacer-sm;
  margin-bottom: $spacer-xxsm;
}

.error {
  display: block;
  font-size: $font-size-sm;
  color: $red;
}

.tags {
  margin: $font-size-sm 0;
  display: flex;
}

.tagContainer {
  display: flex;
  margin-right: $spacer-sm;
  padding: $spacer-xxsm $spacer-xsm;
  border-radius: px-to-rem(4px);
  background-color: $light-bg;
  color: $text-color;
  transition: all 0.2s;

  &:hover {
    background-color: darken($light-bg, 5);
  }
}

.name {
  margin-right: $spacer-xsm;
}

.remove {
  display: flex;
  align-self: center;
  cursor: pointer;
}

.removeIcon {
  stroke-width: 2;
}
