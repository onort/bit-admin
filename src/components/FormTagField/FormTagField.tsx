import React from "react"
import { FieldProps } from "formik"
import cx from "classnames"

import styles from "./FormTagField.scss"
import { Tag } from "./"
import { ErrorMessage } from "../"
import { TagData } from "../../types"

// Custom component props are untyped as of formik v1.3.1
interface Props extends FieldProps {
  label?: string
  required?: boolean
  tagsarrayname: string
  tags: TagData[]
}

const FormTagField: React.SFC<Props> = ({
  form,
  field,
  tagsarrayname,
  ...props
}) => {
  const { errors, touched, values, setFieldValue } = form
  const { name, value } = field
  const handleTagAdd: React.KeyboardEventHandler = (
    e: React.KeyboardEvent<HTMLInputElement>
  ) => {
    if (e.key === "Enter") {
      e.preventDefault()
      values[tagsarrayname].push({ id: value.trim(), name: value.trim() })
      setFieldValue(tagsarrayname, values[tagsarrayname])
      setFieldValue(name, "")
    }
  }
  const handleRemove = (tagId: string) => {
    const newArr = values[tagsarrayname].filter((t: TagData) => t.id !== tagId)
    setFieldValue(tagsarrayname, newArr)
  }
  const labelClassName = cx(styles.label, { [styles.required]: props.required })
  return (
    <div className={styles.container}>
      {props.label && (
        <label className={labelClassName} htmlFor={name}>
          {props.label}
        </label>
      )}
      <input
        className={styles.input}
        {...field}
        {...props}
        onKeyPress={handleTagAdd}
      />
      {values[tagsarrayname].length > 0 && (
        <div className={styles.tags}>
          {values[tagsarrayname].map((tag: TagData) => (
            <Tag key={tag.id} tag={tag} onRemove={handleRemove} />
          ))}
        </div>
      )}
      {touched[name] &&
        !!errors[name] && <ErrorMessage small={true} message={errors[name]} />}
      {touched[name] &&
        !!errors[tagsarrayname] && (
          <ErrorMessage small={true} message={errors[tagsarrayname]} />
        )}
    </div>
  )
}

FormTagField.defaultProps = {
  required: false
}

export default FormTagField
