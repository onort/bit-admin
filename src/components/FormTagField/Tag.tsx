import React from "react"
import { MdClose as CloseIcon } from "react-icons/md"

import styles from "./FormTagField.scss"
import { Tag as TagType } from "../../types"

interface Props {
  tag: TagType
  onRemove: (id: string) => void
}

const Tag: React.SFC<Props> = props => {
  const { id, name } = props.tag
  const handleRemove = () => props.onRemove(id)
  return (
    <div className={styles.tagContainer}>
      <span className={styles.name}>{name}</span>
      <span className={styles.remove} onClick={handleRemove}>
        <CloseIcon className={styles.removeIcon} />
      </span>
    </div>
  )
}

export default Tag
