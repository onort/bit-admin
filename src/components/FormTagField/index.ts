import FormTagField, { TagType } from "./FormTagField"
import Tag from "./Tag"

export default FormTagField

export { TagType, Tag }
