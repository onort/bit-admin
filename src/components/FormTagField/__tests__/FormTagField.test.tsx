import React from "react"
import { mount, shallow } from "enzyme"
import { Formik, Form, Field } from "formik"

import FormTagField, { Tag } from "../"

// TODO: Mock Apollo
describe("<FormTagField />", () => {
  it.skip("should match snapshot", () => {
    const testValues = { test: "", testArray: [] }
    const fakeSubmit = jest.fn()
    const wrapper = shallow(
      <Formik initialValues={testValues} onSubmit={fakeSubmit}>
        {props => {
          return (
            <form onSubmit={props.handleSubmit}>
              <Field
                name="test"
                tagsarrayname="testArray"
                label="Tag Test"
                component={FormTagField}
              />
            </form>
          )
        }}
      </Formik>
    )
    expect(wrapper).toMatchSnapshot()
    expect(wrapper.html()).toMatchSnapshot()
  })

  it.skip("should add a tag", () => {
    const testValues = { test: "", testArray: [] }
    const fakeSubmit = jest.fn()
    const wrapper = mount(
      <Formik initialValues={testValues} onSubmit={fakeSubmit}>
        {props => {
          return (
            <Form {...props}>
              <Field
                name="test"
                tagsarrayname="testArray"
                label="Tag Test"
                component={FormTagField}
              />
            </Form>
          )
        }}
      </Formik>
    )
    wrapper
      .find("input.input")
      .simulate("change", { target: { name: "test", value: "testTag" } })
    wrapper.simulate("keyPress", { key: "Enter" })
    expect(wrapper.find(Tag).exists()).toBe(true)
    wrapper.unmount()
  })
})
