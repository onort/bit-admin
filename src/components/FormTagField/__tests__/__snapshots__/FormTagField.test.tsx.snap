// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<FormTagField /> should match snapshot 1`] = `
<ContextProvider
  value={
    Object {
      "dirty": false,
      "errors": Object {},
      "handleBlur": [Function],
      "handleChange": [Function],
      "handleReset": [Function],
      "handleSubmit": [Function],
      "initialValues": Object {
        "test": "",
        "testArray": Array [],
      },
      "isSubmitting": false,
      "isValid": false,
      "isValidating": false,
      "registerField": [Function],
      "resetForm": [Function],
      "setError": [Function],
      "setErrors": [Function],
      "setFieldError": [Function],
      "setFieldTouched": [Function],
      "setFieldValue": [Function],
      "setFormikState": [Function],
      "setStatus": [Function],
      "setSubmitting": [Function],
      "setTouched": [Function],
      "setValues": [Function],
      "submitCount": 0,
      "submitForm": [Function],
      "touched": Object {},
      "unregisterField": [Function],
      "validate": undefined,
      "validateField": [Function],
      "validateForm": [Function],
      "validateOnBlur": true,
      "validateOnChange": true,
      "validationSchema": undefined,
      "values": Object {
        "test": "",
        "testArray": Array [],
      },
    }
  }
>
  <form
    onSubmit={[Function]}
  >
    <C
      component={[Function]}
      label="Tag Test"
      name="test"
      tagsarrayname="testArray"
    />
  </form>
</ContextProvider>
`;

exports[`<FormTagField /> should match snapshot 2`] = `"<form><div class=\\"container\\"><label class=\\"label\\" for=\\"test\\">Tag Test</label><input class=\\"input\\" value=\\"\\" name=\\"test\\" label=\\"Tag Test\\"/></div></form>"`;
